### PR TITLE
Creating functions checking only diffs not whole files

### PR DESCRIPTION
**Pull request by eiditakahashi**
@melloflavio is this the way you were thinking? Or did I totally misunderstood the idea?

### CODE REVIEW

**Comment by melloflavio** [comment]
@eiditakahashi pretty much in that direction.

The original structure i though of was:
- Define each check as a single function, which receives as input the necessary data (e.g.: `file`, `line`, `diffForFile`, etc).
- Go over the whole diff/file only once.
  - Inside the loop you'd conditionally call the `checkLineForPlatform` function.
  - Each platform strategy would be a composition of the checks defined earlier

Something along the lines of:

```ruby
# Beginning Danger
checkForCommonPRRelatedStuff()
checkForCommonChangedFilesStuff()

modified_files.each do |file|
  begin 
    checkForFile(file)
  rescue
    message "Could not read file #{file}, does it really exist?"
  end
end

def checkForFile(file) 
  checkForFileCommon(file)
  switch (platform)
  case: "ios":
    checkForFileIos(file)
    break;
  case: "node":
    checkForFileNode(file)
    break;
  (...)
end

def checkForFileCommon(file)
  checkHttps(file)
end

def checkForFileNode(file)
  checkEngines(file)
  checkYarnLock(file)
  checkWeaklyTypedReturn(file)
end
```

IMO we should be reading every single line of the changed files. A simple regex.scan in the diff should suffice. I'd also implement a helper `checkDiffForInclusion(string)` function to avoid repeating the `\+.*` pattern over and over...
Although I concede there are some special cases where checking every line is desirable. Mainly on sensitive topics which we want to double check every time.

File: Dangerfile [hunk 1]
```diff
@@ -5,90 +5,75 @@ if !defined? @platform
...
 
 def checkForEnginesVersion(file)
   # Keep engines version synced between declarations
   if file == "package.json"
-    fileDiff = git.diff_for_file(file)
+    # fileDiff = git.diff_for_file(file)
     engines = ["node","npm","yarn"]
     engines.each do |engine|
-      engineVersionMatches = fileDiff.patch.scan(/\+.*(\"#{engine}\")/)
+      engineVersionMatches = checkForRegex(file, /\+.*(\"#{engine}\")/)
```
Comment (felipesabino) [inline Dangerfile:84]: [question] Is this fixing this false positive? https://github.com/indigotech/br-pro120-jogo_das_forcas-server-api/pull/14#issuecomment-289065987

File: Dangerfile [hunk 2]
```diff
@@ -98,17 +83,19 @@ def checkForEnginesVersion(file)
...
+  returnAnyMatches.each do |returnAnyMatch|
+    warn("Possibly returning 'any' in a function, prefer having a strongly typed return. `#{file}` at line `#{returnAnyMatch}`.") if returnAnyMatch
+  end
 end
 
 def checkForNpmInstallGlobal(file)
   # Warn developers that they are not supposed to use this flag
-  fileDiff = git.diff_for_file(file)
-  npmInstallMatches = fileDiff.patch.scan(/\+.*npm install (-g|--global)/)
+  npmInstallMatches = checkForRegex(file, /\+.*npm install (-g|--global)/)
```
Comment (felipesabino) [inline Dangerfile]: can we ignore certain files?

some PRs are _failing_ due to `npm install -g yarn` instruction on `README`, ex: https://github.com/indigotech/br-pro120-jogo_das_forcas-server-api/pull/10#issuecomment-287921670

File: Dangerfile [hunk 3]
```diff
@@ -215,97 +252,103 @@ if @platform == "ios"
...
-    end
-  end
+validatePackageJson(modified_files, github.pr_diff) if @platform == "nodejs" || @platform == "web"
 
-  validatePackageJson(modified_files, github.pr_diff)
+modified_files.each do |file|
+  begin
+    checkForFile(file)
+  rescue
+    message "Could not read file #{file}, does it really exist?"
```
Comment (felipesabino) [inline Dangerfile]: Can we check if file was excluded or moved before raising this error?

Most of my PRs have a long list with this kind of message that is just noise and can be completely ignored by devs. example: https://github.com/indigotech/br-pro120-jogo_das_forcas-server-api/pull/25#issuecomment-290737206

**Review by felipesabino — APPROVED** [review]

File: Dangerfile (hunk 1 above)
Comment (eiditakahashi) [reply Dangerfile:84]: Which one is a false positive in this PR? Or you are talking about the `missing files` messages? I tried to run danger locally but this commit is already gone and can't be found :pensive:. I tried in another PR https://github.com/indigotech/br-pro120-jogo_das_forcas-server-api/pull/25

Result with latest release (1.1.3):
```
Messages:
- [ ] Could not read file src/api/admin/card.test.ts, does it really exist?
- [ ] Could not read file src/api/admin/context.test.ts, does it really exist?
- [ ] Could not read file src/api/admin/license.test.ts, does it really exist?
- [ ] Could not read file src/api/admin/registration.test.ts, does it really exist?
- [ ] Could not read file src/api/admin/user.test.ts, does it really exist?
- [ ] Could not read file src/api/card.test.ts, does it really exist?
- [ ] Could not read file src/api/context.test.ts, does it really exist?
- [ ] Could not read file src/domain/card-crud.use-case.ts, does it really exist?
- [ ] Could not read file src/domain/card-list.use-case.ts, does it really exist?
- [ ] Could not read file src/test/index.ts, does it really exist?
```
Result with this branch:
```
Warnings:
- [ ] Possibly returning 'any' in a function, prefer having a strongly typed return. `src/api/card.test.ts` at line `+  async register(app: any): Promise<any>`.
- [ ] Possibly returning 'any' in a function, prefer having a strongly typed return. `src/api/context.test.ts` at line `+  async register(app: any): Promise<any>`.
- [ ] Possibly returning 'any' in a function, prefer having a strongly typed return. `src/domain/card-list.use-case.ts` at line `+  async execute(pagination: PaginationModel, request: CardListRequest): Promise<any>`.
```

File: Dangerfile (hunk 3 above)
Comment (eiditakahashi) [reply Dangerfile]: On the latest release (1.1.3) there is a bug in a function used only for files with extension `.ts` that is causing these messages flood:
```
def checkForWeaklyTypedFunctionReturn(line)
  warn("Possibly returning 'any' in a function, prefer having a strongly typed return. `#{file}` at line `#{line}`.") if line =~ /<any>/im
end
```
It is trying to print var `file` but it doesn't exist in this context :pensive:

This PR already contains a correction for this (check my reply in your comment above) :point_up:

But I did a function to check if file exists as well :satisfied:

**Pull request merged**